### DISCUSSION TITLE
Accelerate decompilation process (to get usings only)

### DISCUSSION THREAD

**Discussion by JiriOgurek** [icsharpcode/ILSpy#2977] [Help] [answered]
Hi, I would like to ask you, if there's an option to accelerate decompilation process to get usings only, please?

**Comment by dgrunwald** [comment]
Usings depend on whether the decompiler had to introduce temporary variables (in which case it'll spell out the variable type which might need a using), or could simplify the logic into a single expression (in which case the type of the temporary doesn't need a using).
There's no way to calculate the usings without doing a full decompilation.

**Reply by JiriOgurek** [reply]
@dgrunwald, thank you for your kind reply. If there is full decompilation needed, is there any general option to accelerate the process (not affecting the usings), please?

**Reply by dgrunwald** [reply]
No. Note that basically every option the decompiler has, has the potential to change which temporaries end up becoming varaibles, and thus also changes which `using`s are generated.
I'm not really sure why you think the `using`s are in any way interesting -- they're really an artifact of the decompilation process, and don't directly correspond to anything in the program (e.g. you can have a dependency on a type without `using` its namespace).

**Reply by JiriOgurek** [reply]
@dgrunwald, thank you for an explanation, I have created vulnerability scanner, based on reporting suspicious refrences (like log4net), so I am interested in getting all references as fast as possible, this is my goal. Would you recommend other approach to extract the references instead of getting usings, please?

**Reply by siegfriedpammer** [reply]
You could do something similar to https://github.com/icsharpcode/ILSpy/blob/master/ICSharpCode.Decompiler/CSharp/RequiredNamespaceCollector.cs although it might not be 100% accurate, it creates an over-approximation of the used namespaces of all types and members.

**Reply by JiriOgurek** [reply]
@siegfriedpammer, thank you for the hint :-) So you would prefer getting the usings from full decompilation as it's more accurate?

**Reply by dgrunwald** [reply]
RequiredNamespaceCollector will give you all namespaces for types or methods that are used in the IL; the `using`s from the full decompilation will only give you the namespaces for the types that appear in the decompiled sourcecode.

So if you have a call chain `a.Foo().Bar();`, then the RequiredNamespaceCollector will report the namespace for `Foo`'s return type (where `Bar` is declared), even though there won't be a `using` for that.

Which variant is "more accurate" depends on what you want -- for a vulnerability scanner, I would expect that all uses of vulnerable classes are reported, even if they don't involve a literal `using`.

**Reply by JiriOgurek** [reply]
The scanner is based on scanning for suspicious (reported) nugets, that's why I am interested in usings /references/, for this task full decompilation is "more accurate", right? I am not interested in class level analysis, this has to be done manually afterwards as it's not easy to do.

**Reply by christophwille** [reply]
Wouldn't looking for references be sufficient? Or do you have the problem of transitive references?

**Reply by JiriOgurek** [reply]
I find working with usings more reliable, as it's required for existing code, references are tricky - can be unused, have dependencies, ... Correct me, if I am wrong, please. Anyway is there a way how to get references, please?

**Reply by dgrunwald** [reply, marked as answer]
AFAIK the C# compiler already ignores unused references (doesn't store them in the .dll), so if you don't want class-level analysis, references from the .dll should be good enough. (though note that references use assembly-granularity, where as usings are namespace-granularity)
You don't need a decompiler at all in that case; you can e.g. directly use System.Reflection.Metadata, or maybe Mono.Cecil as that tends to be easier to use.

**Reply by JiriOgurek** [reply]
@dgrunwald, thank you for GREAT advice, it's lightning FAST!
